### PR TITLE
supporting static report for id stitcher

### PR DESCRIPTION
**Pull request by joker2411**
## Description of the change

> Ticket [Link](https://linear.app/rudderstack/issue/PRML-1000/add-llm-analysis-of-static-report).

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/predictions/profiles_mlcorelib/py_native/id_stitcher/config.py [hunk 1]
```diff
@@ -1 +1,2 @@
 LLM_SERVICE_URL = "https://api.rudderstack.com/v2/profiles/idStitcher/debug"
+LLM_DIRECT_REQUEST_URL = "https://api.rudderstack.com/v2/invoke"
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/py_native/id_stitcher/config.py:2]: Let's name both as `LLM_ID_DEBUG_URL` and `LLM_INVOKE_URL`

**Review by dpatchigolla — APPROVED** [review]
added minor rename suggestion. changes look good.
Please take a signoff from Shashank too

File: src/predictions/profiles_mlcorelib/py_native/id_stitcher/table_report.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def get_total_main_ids(self) -> int:
         entity_key = self.entity["IdColumnName"]
         query = f"select count(distinct {entity_key}) as count from {self.db}.{self.schema}.{self.output_table}"
         result = run_query(self.wh_client, query)
-        return result["COUNT"][0]
+        return 0 if result.empty else int(result["COUNT"][0])
```
Comment (shekhar-rudder) [inline src/predictions/profiles_mlcorelib/py_native/id_stitcher/table_report.py:110]: Typecasting should be done in `llm_report.py` for consistency

File: src/predictions/profiles_mlcorelib/py_native/id_stitcher/table_report.py [hunk 3]
```diff
@@ -158,7 +158,7 @@ def get_average_edge_count(self, id_type: str = None) -> float:
         )
         """
         result = run_query(self.wh_client, query)
-        return 0 if result.empty else result["AVG_EDGE_COUNT"][0]
+        return 0 if result.empty else float(result["AVG_EDGE_COUNT"][0])
```
Comment (shekhar-rudder) [inline src/predictions/profiles_mlcorelib/py_native/id_stitcher/table_report.py:161]: Same here

File: src/predictions/profiles_mlcorelib/py_native/id_stitcher/llm_report.py [hunk 4]
```diff
@@ -90,28 +133,19 @@ def _get_report(self, report):
...
-            "prompt": prompt,
-            "session_id": self.session_id,
-            "tables": {
-                "edges": self.table_report.edges_table,
-                "id_graph": self.table_report.output_table,
-            },
-            "warehouse_credentials": self.warehouse_credentials,
-            "report": self._get_report(self.table_report.analysis_results),
-        }
+    def _request(self, URL, body):
```
Comment (shekhar-rudder) [inline src/predictions/profiles_mlcorelib/py_native/id_stitcher/llm_report.py:143]: URL -> url

File: src/predictions/profiles_mlcorelib/py_native/id_stitcher/llm_report.py [hunk 5]
```diff
@@ -46,22 +47,64 @@ def run(self):
         if not consent:
             print("LLM-based analysis is disabled. Skipping LLM-based analysis.")
             return
+
+        print("\n\nGenerating id_stitcher analysis:")
+        self._interpret_results_with_llm()
```
Comment (shekhar-rudder) [inline src/predictions/profiles_mlcorelib/py_native/id_stitcher/llm_report.py:52]: What happens to this API call if the site config doesn't have access token ?

File: src/predictions/profiles_mlcorelib/py_native/id_stitcher/config.py [hunk 6]
```diff
@@ -1 +1,2 @@
-LLM_SERVICE_URL = "https://api.rudderstack.com/v2/profiles/idStitcher/debug"
+LLM_ID_DEBUG_URL = "https://api.rudderstack.com/v2/profiles/idStitcher/debug"
+LLM_INVOKE_URL = "https://api.rudderstack.com/v2/invoke"
```
Comment (shekhar-rudder) [inline src/predictions/profiles_mlcorelib/py_native/id_stitcher/config.py:2]: Extract out the common url `https://api.rudderstack.com/v2` in a constant `BASE_URL`

Comment (shekhar-rudder) [reply src/predictions/profiles_mlcorelib/py_native/id_stitcher/config.py:2]: https://api.rudderstack.com/v2/invoke -> https://api.rudderstack.com/v2/llm/invoke

Comment (joker2411) [reply src/predictions/profiles_mlcorelib/py_native/id_stitcher/config.py]: changed.

Also made the same change to API name on server side.

File: src/predictions/profiles_mlcorelib/py_native/id_stitcher/llm_report.py (hunk 5 above)
Comment (joker2411) [reply src/predictions/profiles_mlcorelib/py_native/id_stitcher/llm_report.py:53]: got that.
But my question was, when client is asked to enter a access token, is it supposed to be the one fetched from his RudderStack account or they can enter any string ?

Comment (shekhar-rudder) [reply src/predictions/profiles_mlcorelib/py_native/id_stitcher/llm_report.py:53]: It is supposed to be a valid access token. So yes, it has to be from a RudderStack account.

**Review by shekhar-rudder — APPROVED** [review]

**Pull request merged**